docs/design/ha_master.md

```diff
@@ -0,0 +1,259 @@
+
+When generating command line for etcd we will always assume it’s part of a cluster
+(initially of size 1) and list all VMs for kubernetes masters (see VM naming section).
+Based on that, we will set the following flags:
+* `-initial-cluster` - list of all hostnames/DNS names for master machines
+* `-initial-cluster-state` (keep in mind that we are adding masters one by one):
+  * `new` if we are adding the first replica, i.e. the list of existing masters is empty
+  * `existing` if there are more than one replica, i.e. the list of existing masters is non-empty.
+
+This will allow us to have exactly the same logic for HA and non-HA master. List of DNS names for VMs
+with master replicas will be generated in `kube-up.sh` script and passed to as a env variable
+`INITIAL_ETCD_CLUSTER`.
+
+### apiservers
+
+All apiservers will work independently. They will contact etcd on 127.0.0.1, i.e. they will always contact
+etcd replica running on the same VM. If needed, such requests will be forwarded by etcd server to the
+etcd leader. This functionality is completely hidden from the client (apiserver
+in our case).
+
+Caching mechanism, which is implemented in apiserver, will not be affected by
+replicating master because:
+* GET requests go directly to etcd
+* LIST requests go either directly to etcd or to cache populated via watch
+  (depending on the ResourceVersion in ListOptions). In the second scenario,
+  after a PUT/POST request, changes might not be visible in LIST response.
+  This is however not worse than it is with the current single master.
+* WATCH does not give any guarantees when change will be delivered.
+
+#### DNS & load balancer
+
+In order to expose a single endpoint for the user, we will introduce a DNS name for the cluster.
+In a single-homed master it will point directly to the apiserver. In multi-homed
+scenario we will have two options:
+
+* create a load balancer in front of all apiservers and update DNS name appropriately
+* use round-robin DNS technique to access all apiservers
+
+The advantage from the client perspective of introducing a DNS name/load-balancer is that
+it will be transparent how many apiservers we have.
+
+#### Certificates
+
+Certificates used by apiserver, which are generated during cluster creation, will
+include IP address of the master VM and DNS name for the cluster. That way old
+clients will not need to be changed, however they will still use IP, not DNS name.
+As a result, if administrator adds more master replicas, old clients will not take
+advantage of this. To fix this clients will have to *manually* change cluster
+address in their configuration.
+
+To mitigate the need to manually update multiple clients we could also take the
+initial IP for the master VM and make it a virtual one and point it to all the
+replicas. This, however, will not be automated.
+
+As a result of changed certificates, talking directly to apiserver (except for the
```

Comment:
What do you mean by certificates API? Are you saying we should regenerate certificates with each instance change? Do you think we can deliver this in v2 of this feature or would be a blocker? I'm happy to add this as a future improvement. WDYT?